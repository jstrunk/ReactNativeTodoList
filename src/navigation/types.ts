import { StackNavigationProp } from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";

export type RootStackParamList = {
  TodoSelect: undefined;
  TodoList: {id: string};
};

export type TodoSelectScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoSelect'
>;

export type TodoListScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoList'
>;

export type TodoListScreenRouteProp = RouteProp<
  RootStackParamList,
  'TodoList'
>;

