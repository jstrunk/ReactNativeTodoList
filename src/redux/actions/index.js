import { ADD_ITEM, DELETE_ITEM, TOGGLE_ITEM, DELETE_COMPLETED } from '../constants';

export function addItem(text) {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index) {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index) {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted() {
  return { type: DELETE_COMPLETED };
}
