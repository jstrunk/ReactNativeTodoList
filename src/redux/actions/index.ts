import { ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  DELETE_COMPLETED,
  TodoActionType, } from '../types';

export function addItem(text: string): TodoActionType {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index: number): TodoActionType {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index: number): TodoActionType {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted(): TodoActionType {
  return { type: DELETE_COMPLETED };
}
