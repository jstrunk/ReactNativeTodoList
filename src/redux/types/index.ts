export interface Todo {
  key: string;
  done: boolean;
}

export interface ITodoList {
  byId: Record<string,Todo>;
  allItems: Array<string>;
}

export interface ITodoState {
  todoList: ITodoList;
}

export const defaultState: ITodoList = {
  byId: {},
  allItems: [],
};

export const ADD_ITEM = "ADD_ITEM";
export const TOGGLE_ITEM = "TOGGLE_ITEM";
export const DELETE_ITEM = "DELETE_ITEM";
export const DELETE_COMPLETED = "DELETE_COMPLETED";

interface AddItemAction {
  type: typeof ADD_ITEM;
  text: string;
  id: string;
}

interface ToggleItemAction {
  type: typeof TOGGLE_ITEM;
  id: string;
}

interface DeleteItemAction {
  type: typeof DELETE_ITEM;
  id: string;
}

interface DeleteCompletedAction {
  type: typeof DELETE_COMPLETED;
  ids: Array<string>;
}

export type TodoActionType = AddItemAction | ToggleItemAction |
  DeleteItemAction | DeleteCompletedAction;
