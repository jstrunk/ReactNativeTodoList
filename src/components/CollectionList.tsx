import React, {Component} from 'react';
import {FlatList} from 'react-native';
import { ITodoIndex } from '../redux/types';
import { TodoSelectScreenNavigationProp } from '../navigation/types';
import styles from '../styles';
import { ListItem } from 'react-native-elements';

interface Props {
  items: Array<ITodoIndex>;
  loadList: (id: string) => void;
  navigation: TodoSelectScreenNavigationProp;
}

interface ItemProps {
  item: ITodoIndex;
}

export default class CollectionList extends Component<Props> {
  pressList(id: string): void {
    console.log('navigating to ' + id);
    this.props.navigation.navigate('TodoList', {id});
  }

  renderItem = (itemProps: ItemProps): JSX.Element => (
    <ListItem titleStyle={styles.item} onPress={() => this.pressList(itemProps.item.id)} title={itemProps.item.name} />
  ); 

  render(): JSX.Element {
    return (
      <FlatList
        style={styles.list}
        data={this.props.items}
        renderItem={this.renderItem}
        keyExtractor={(item: ITodoIndex) => item.id }
      />
    );
  }
}
