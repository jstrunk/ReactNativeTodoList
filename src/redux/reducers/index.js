import { combineReducers } from 'redux';
import listReducer from '../reducers/listReducer';

export default combineReducers(
  { todoList: listReducer }
);