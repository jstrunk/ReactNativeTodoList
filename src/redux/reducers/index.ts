import { combineReducers } from 'redux';
import listReducer from '../reducers/listReducer';
import itemReducer from '../reducers/itemReducer';
import { ITodoState } from '../types';

export default combineReducers<ITodoState>(
  {
    todoList: listReducer,
    todoItems: itemReducer,
  }
);
