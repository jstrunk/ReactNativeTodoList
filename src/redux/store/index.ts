import AsyncStorage from '@react-native-community/async-storage';
import { createStore, Store } from 'redux';
import { persistStore, persistReducer, Persistor } from 'redux-persist';
import rootReducer from '../reducers';

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

export interface PersistStore {
  store: Store,
  persistor: Persistor,
}
const configureStore = (): PersistStore => {
  const store = createStore(persistedReducer);
  const persistor = persistStore(store);
  return { store, persistor };
}

export default configureStore;
