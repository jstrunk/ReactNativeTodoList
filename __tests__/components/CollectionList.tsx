import 'react-native';
import React from 'react';
import CollectionList from '../../src/components/CollectionList';
import { render, fireEvent } from '@testing-library/react-native';
import { ITodoIndex } from '../../src/redux/types';

describe('CollectionList component', () => {
  const items: Array<ITodoIndex> = [{id: 'list1', name: 'Test List1'}, {id: 'list2', name: 'Test List2'}];
  let loadList: jest.Mock<any,any>;
  const navigation: Record<string,unknown> = {};

  beforeEach(() => {
    loadList = jest.fn();
    
    navigation.navigate = jest.fn();
  });

  test('renders correctly', () => {
    const { toJSON } = render(
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore: ts(2769) for incorrect type of mocked navigation
      <CollectionList items={items} loadList={loadList} navigation={navigation} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('has items', () => {
    const { getAllByText } = render(
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore: ts(2769) for incorrect type of mocked navigation
      <CollectionList items={items} loadList={loadList} navigation={navigation} />
    );

    const fooElements = getAllByText('Test List2');
    expect(fooElements).toHaveLength(1);
  });

  it('fires a navigation event on press', () => {
    const { getByText } = render(
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore: ts(2769) for incorrect type of mocked navigation
      <CollectionList items={items} loadList={loadList} navigation={navigation} />
    );

    const fooElement = getByText('Test List2');
    
    fireEvent.press(fooElement);

    expect(navigation.navigate).toHaveBeenCalledTimes(1);
    expect(navigation.navigate).toHaveBeenCalledWith('TodoList', {id: 'list2'});
  });
});