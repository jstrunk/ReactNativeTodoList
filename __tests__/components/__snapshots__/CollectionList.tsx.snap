// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionList component renders correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "id": "list1",
        "name": "Test List1",
      },
      Object {
        "id": "list2",
        "name": "Test List2",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "marginBottom": 40,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderColor": "#bcbbc1",
              "flexDirection": "row",
              "padding": 14,
            }
          }
          testID="padView"
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "black",
                  "fontSize": 17,
                  "textAlign": "left",
                }
              }
              testID="listItemTitle"
              theme={
                Object {
                  "colors": Object {
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "primary": "#007aff",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                }
              }
            >
              Test List1
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderColor": "#bcbbc1",
              "flexDirection": "row",
              "padding": 14,
            }
          }
          testID="padView"
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "black",
                  "fontSize": 17,
                  "textAlign": "left",
                }
              }
              testID="listItemTitle"
              theme={
                Object {
                  "colors": Object {
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "primary": "#007aff",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                }
              }
            >
              Test List2
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
